GDJS/Runtime/variable.js

```diff
@@ -20,278 +26,359 @@
+  return this._str;
 };
 
 /**
  * Change the value of the variable, considered as a string
- * @method
  * @param {string} newValue The new string to be set
  */
-gdjs.Variable.prototype.setString = function(newValue) {
-	this._str = newValue;
-	this._numberDirty = true;
-	this._stringDirty = false;
+gdjs.Variable.prototype.setString = function (newValue) {
+  this._type = 'string';
+  this._str = newValue;
+};
+
+/**
+ * Get the value of the variable, considered as a boolean
+ * @return {boolean}
+ */
+gdjs.Variable.prototype.getAsBoolean = function () {
+  if (this._type !== 'boolean') {
+    if (this._type === 'number') return this._value !== 0;
+    else if (this._type === 'string')
+      return this._str !== '0' && this._str !== '' && this._str !== 'false';
+  }
+
+  return this._bool;
+};
+
+/**
+ * Change the value of the variable, considered as a boolean
+ * @param {boolean} newValue
+ */
+gdjs.Variable.prototype.setBoolean = function (newValue) {
+  this._type = 'boolean';
+  this._bool = newValue;
 };
 
 /**
  * Return true if the variable is a structure.
- * @method
  * @return {boolean} true if the variable is a structure.
+ * @deprecated Use {@link gdjs.Variable.getType} instead.
  */
-gdjs.Variable.prototype.isStructure = function() {
-	return this._isStructure;
+gdjs.Variable.prototype.isStructure = function () {
+  return this._type === 'structure';
 };
 
 /**
  * Return true if the variable is a number.
- * @method
  * @return {boolean} true if the variable is a number.
+ * @deprecated Use {@link gdjs.Variable.getType} instead.
+ */
+gdjs.Variable.prototype.isNumber = function () {
+  return this._type === 'number';
+};
+
+/**
+ * Return true if the variable type is a primitive.
+ * @return {boolean}
+ */
+gdjs.Variable.prototype.isPrimitive = function () {
+  return gdjs.Variable.isPrimitive(this._type);
+};
+
+/**
+ * Returns the type of the variable.
+ * @return {VariableType}
  */
-gdjs.Variable.prototype.isNumber = function() {
-	return !this._isStructure && !this._numberDirty;
+gdjs.Variable.prototype.getType = function () {
+  return this._type;
 };
 
 /**
  * Return the object containing all the children of the variable
- * @method
  * @return {Object.<string, gdjs.Variable>} All the children of the variable
  */
-gdjs.Variable.prototype.getAllChildren = function() {
-	return this._children;
+gdjs.Variable.prototype.getAllChildren = function () {
+  return this._children;
+};
+
+/**
+ * Return an Array containing all the children of the variable
+ * @return {gdjs.Variable[]}
+ */
+gdjs.Variable.prototype.getAllChildrenList = function () {
+  return this._type === 'structure'
+    ? Object.values(this._children)
+    : this._type === 'array'
+    ? this._childrenList
+    : [];
+};
+
+/**
+ * Return the length of the
+ * @return {number}
+ */
+gdjs.Variable.prototype.getChildrenCount = function () {
+  if (this.isPrimitive()) return 0;
+  //@ts-ignore This cannot return null as type isn't a primitive.
```

Comment:
Yes I already fixed this locally, I returned null instead of an empty array before but I had to change that to make the for each child event not crash.

Reply:
Sounds dangerous then, it means that the typing is false? You should never ever get null?